### PR TITLE
Moved functions to pytorch_utils.py

### LINKED ISSUE #15543: Move generic PyTorch utils function from modeling_utils.py to pytorch_utils
# 🚀 Feature request

> Since we're creating a new module file, can we maybe move some other functions there? Like the no_init_weight context manager, and:
> 
> all pruning stuff
> apply_chunking_to_forward
> get_parameter_device
> get_parameter_dtype
> so that the modeling utils file stays focused on PreTrainedModel and the layers it defines?

Taken from comment here: https://github.com/huggingface/transformers/pull/15498#pullrequestreview-874783288

### PR DESCRIPTION
**Pull request by anmolsjoshi**
# What does this PR do?

Moved all pruning stuff, apply_chunking_to_forward to pytorch_utils.py

Fixes #15543

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten @sgugger

### CODE REVIEW

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:100]: For backward compatibility, these functions should still be imported in this module.

File: examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py
Comment (sgugger) [inline examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py:32]: Research projects use a pinned version of Transformers, so you shouldn't update this file.

File: src/transformers/models/ctrl/modeling_ctrl.py
Comment (sgugger) [inline src/transformers/models/ctrl/modeling_ctrl.py:25]: Should be imported from `pytorch_utils`

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (sgugger) [inline src/transformers/models/decision_transformer/modeling_decision_transformer.py:27]: Should be imported from pytorch_utils

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (sgugger) [inline src/transformers/models/gpt2/modeling_gpt2.py:36]: Should be imported from pytorch_utils.

File: src/transformers/models/imagegpt/modeling_imagegpt.py
Comment (sgugger) [inline src/transformers/models/imagegpt/modeling_imagegpt.py:35]: Should be imported from pytorch_utils.

File: src/transformers/models/openai/modeling_openai.py
Comment (sgugger) [inline src/transformers/models/openai/modeling_openai.py:29]: Should be imported from pytorch_utils.

**Review by sgugger — COMMENTED** [review]
Thanks for tackling this! There are a few imports that seem wrong in some of the modeling files, could you fix them?

File: examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py
Comment (anmolsjoshi) [reply examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py]: Reverted

File: src/transformers/models/ctrl/modeling_ctrl.py
Comment (anmolsjoshi) [reply src/transformers/models/ctrl/modeling_ctrl.py]: Incorporated

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (anmolsjoshi) [reply src/transformers/models/gpt2/modeling_gpt2.py]: Incorporated

File: src/transformers/models/openai/modeling_openai.py
Comment (anmolsjoshi) [reply src/transformers/models/openai/modeling_openai.py]: Incorporated

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py
Comment (anmolsjoshi) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py]: Incorporated

File: src/transformers/modeling_utils.py
Comment (anmolsjoshi) [reply src/transformers/modeling_utils.py:100]: @sgugger does that mean a copy of the functions should remain in modeling_utils.py?

Comment (anmolsjoshi) [reply src/transformers/modeling_utils.py:100]: Or as described [here](https://stackoverflow.com/a/57850909)?

Comment (sgugger) [reply src/transformers/modeling_utils.py:100]: No, just `from .pytorch_utils import find_pruneable_heads_and_indices` (and same for the others)

**Review by patrickvonplaten — APPROVED** [review]
Went through all the files and it looks good to me! 

Think it would make a bit more sense to leave `Conv1D` in `modeling_utils.py` next to classes like `PoolerAnswerClass` instead of moving it to a pytorch utils class since it's more related to modeling IMO - what do you think @sgugger ?

**Comment by sgugger** [comment]
Those should be removed entirely honestly (components should be in model files, not modeling_utils), so I care little where they are :-)

**Comment by anmolsjoshi** [comment]
@patrickvonplaten won't moving Conv1D back to modeling_utils.py cause circular imports?

**Comment by anmolsjoshi** [comment]
@patrickvonplaten any thoughts?

**Comment by patrickvonplaten** [comment]
Good to go for me!

**Comment by patrickvonplaten** [comment]
Waiting for final review from @sgugger

**Comment by sgugger** [comment]
I'm good on my side, was waiting for you Patrick ;-)

**Pull request merged**
